File: sol/solution/0001-0100/0039. Combination Sum/Solution.kt
class Solution {
    fun combinationSum(candidates: IntArray, target: Int): List<List<Int>> {
        candidates.sort()
        val result = mutableListOf<List<Int>>()
        val current = mutableListOf<Int>()
        fun backtrack(start: Int, remaining: Int) {
            if (remaining == 0) {
                result.add(ArrayList(current))
                return
            }
            for (i in start until candidates.size) {
                val num = candidates[i]
                if (num > remaining) break  // prune
                current.add(num)
                backtrack(i, remaining - num)  // reuse same candidate
                current.removeAt(current.size - 1)
            }
        }
        backtrack(0, target)
        return result
    }
}

File: sol/solution/0001-0100/0039. Combination Sum/Solution.scala
object Solution {
    def combinationSum(candidates: Array[Int], target: Int): List[List[Int]] = {
        def backtrack(remain: Int, start: Int, current: List[Int]): List[List[Int]] = {
            if (remain == 0) List(current)
            else if (remain < 0) Nil
            else {
                (start until candidates.length).flatMap { i =>
                    backtrack(remain - candidates(i), i, current :+ candidates(i))
                }.toList
            }
        }
        backtrack(target, 0, Nil)
    }
}